systemd-hosts.d.service

```diff
@@ -3,13 +3,66 @@ Description=Keep your hosts records in logically separated files
 After=local-fs.target
 
 [Service]
+# Security
+ProtectSystem=yes
+ProtectHome=yes
+PrivateDevices=yes
+ProtectKernelTunables=yes
+ProtectKernelModules=yes
+ProtectControlGroups=yes
+ProtectKernelLogs=yes
+SystemCallFilter=@system-service
+SystemCallErrorNumber=EPERM
+NoNewPrivileges=yes
+PrivateTmp=yes
+LockPersonality=yes
+MemoryDenyWriteExecute=yes
+ProtectClock=yes
+RestrictNamespaces=yes
+RestrictRealtime=yes
+RestrictSUIDSGID=yes
+RuntimeDirectory=systemd/hosts.d
+RuntimeDirectoryPreserve=yes
+
+# Service
 Type=oneshot
+
+# Environment
 Environment="HOSTS_FILE=/etc/hosts"
 Environment="HOSTS_DIRECTORY=/etc/hosts.d"
-ExecStart=/bin/sh -c 'echo -e "## Generated by systemd-hosts.d ##\\n" > $HOSTS_FILE'
-ExecStart=/bin/sh -c 'echo -e "## All changes in this file will be ignored ##\\n" >> $HOSTS_FILE'
-ExecStart=/bin/sh -c 'cat $HOSTS_DIRECTORY/*.conf | grep \'^[[:blank:]]*[^[:blank:]#;]\' >> $HOSTS_FILE'
-ExecStart=/bin/chmod ugo=r $HOSTS_FILE
+# Loosely match IPv4 or IPv6 followed by hostname and hostalias(s)
+Environment="HOSTS_REGEX_BASE=^\\([0-9]\\{1,3\\}\\.\\|[0-9a-fA-F]\\{1,4\\}:\\|::\\).\\+\\?\\(\\s\\+[a-zA-Z0-9.-]\\{1,\\}\\)\\{1,\\}$"
+Environment="HOSTS_REGEX_COMMENT=^#.*$"
+# Override with HOSTS_ENABLE_COMMENT=yes to include comments from the source conf files
+Environment="HOSTS_ENABLE_COMMENT=no"
+# Override with HOSTS_ENABLE_BACKUP=no to disable hosts file backup creation
+Environment="HOSTS_ENABLE_BACKUP=yes"
+
+# Condition - validate that matching regex is found in the source before overwriting $HOSTS_FILE
+ExecCondition=/bin/sh -c 'grep "$HOSTS_REGEX_BASE" "$HOSTS_DIRECTORY"/*.conf >/dev/null'
+
+# Create a single backup
+ExecStartPre=/bin/sh -c 'if test "$HOSTS_ENABLE_BACKUP" = "yes"; then cp --dereference -av $HOSTS_FILE ${HOSTS_FILE}~; fi'
+
+# Parse conf files and overwrite $HOSTS_FILE #
+# Maintain $HOSTS_FILE header
+ExecStart=-/bin/sh -c 'echo -e "### Generated by systemd-hosts.d ###\\n" > "$HOSTS_FILE"'
+ExecStart=-/bin/sh -c 'echo "### All changes in this file will be ignored ###" >> "$HOSTS_FILE"'
+
+# Add preceeding newline and accompanying comment to indicate the source conf
+ExecStart=-/bin/sh -c 'find "$HOSTS_DIRECTORY" -maxdepth 1 -type f -iname '*.conf' \
+| sort -n \
+| while read CONF_FILE; \
+do (echo -e "\\n## $CONF_FILE ##"; \
+  HOSTS_REGEX="$HOSTS_REGEX_BASE"; \
+  if test "$HOSTS_ENABLE_COMMENT" = "yes"; then \
+    HOSTS_REGEX="${HOSTS_REGEX_COMMENT}\\|$HOSTS_REGEX"; \
+  fi; \
+  grep -h "$HOSTS_REGEX" \
+"$CONF_FILE") >> "$HOSTS_FILE"; done'
```

Comment:
Oh, coming to think about this, leaving the context for names of retrieved source files indeed seems rather nice.

Although you don't really need the whole `find | sort | read` thing, it can be expressed purely in posix shell with a for loop. Also iirc glob patterns are already undergo alphanumeric sorting, so no need for explicit call to `sort`. And the previous header output also could be composed.

```sh
(
    echo "### Generated by systemd-hosts.d ###"
    echo "### All changes to this file will be ignored ###"
    echo
    for conf_file in "$HOSTS_DIRECTORY"/*.conf; do
        echo "## $conf_file ##"
        cat -- "$conf_file"
        echo
    done
) > "$HOSTS_FILE"
```

Reply:
> Oh, coming to think about this, leaving the context for names of retrieved source files indeed seems rather nice.

I agree, this is extremely helpful since I often use many hosts modifications that can benefit from logical separation.

> Although you don't really need the whole `find | sort | read` thing, it can be expressed purely in posix shell with a for loop. Also iirc glob patterns are already undergo alphanumeric sorting, so no need for explicit call to `sort`. And the previous header output also could be composed.

I explicitly selected a *while* loop instead of a *for* to avoid shell globbing and limitations of `MAX_ARGS`, etc. Use of `sort` is required for this type of piped structure. I think processing a stream makes more sense in most situations. You could argue that you'd never hit some of these issues for most use cases but why not aim for broad support?

Reply:
Frankly, I happened to encounter a max-argument error myself.. nowhere near the production-ready code or anything though. To my knowledge, this limitation doesn't come from the shell itself, but from the kernel trying to pass the cli arguments. So built-in commands should work just fine (just tested it, actually).

And generally, I get to prefer minimize use of external commands (primarily because of the delay arising from the staring new processes, which is quiet noticeable in a loop). And, especially in this case, where the exact same can be expressed much more concisely using only shell built-ins (just compressing the whole piping business into the simple glob).